### PR TITLE
Add engine for unified entrypoints in downstream projects

### PR DESCRIPTION
**Pull request by ZwwWayne**
This PR adds a directory named engine in MMCV, which implements and unifies common functions for starting a training/testing process in downstream codebases. With `mmcv.engine`, downstream codebases like MMAction or MMClassification could directly use these functions and share a similar entry point script for training or testing a model.

This PR does not affect its dependencies' functionality, but its dependencies could update and unify their training/testing scripts after this PR is merged.

### CODE REVIEW

**Comment by xvjiarui** [comment]
`train.py` is an empty file.

File: mmcv/engine/test.py [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+    if rank == 0:
+        prog_bar = mmcv.ProgressBar(len(dataset))
+    time.sleep(2)  # This line can prevent deadlock problem in some cases.
+    for i, data in enumerate(data_loader):
+        with torch.no_grad():
+            result = model(return_loss=False, rescale=True, **data)
+        results.extend(result)
+
+        if rank == 0:
+            batch_size = len(result)
```
Comment (hellock) [inline mmcv/engine/test.py:46]: Is this true for all codebases?

File: mmcv/engine/test.py [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+
+    # collect results from all ranks
+    if gpu_collect:
+        results = collect_results_gpu(results, len(dataset))
+    else:
+        results = collect_results_cpu(results, len(dataset), tmpdir)
+    return results
+
+
+def collect_results_cpu(result_part, size, tmpdir=None):
```
Comment (hellock) [inline mmcv/engine/test.py:58]: Add a docstring.

File: mmcv/engine/test.py [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+        for res in zip(*part_list):
+            ordered_results.extend(list(res))
+        # the dataloader may pad some samples
+        ordered_results = ordered_results[:size]
+        # remove tmp dir
+        shutil.rmtree(tmpdir)
+        return ordered_results
+
+
+def collect_results_gpu(result_part, size):
```
Comment (hellock) [inline mmcv/engine/test.py:101]: Add a docstring.

File: mmcv/utils/config.py [hunk 4]
```diff
@@ -62,6 +62,89 @@ def add_args(parser, cfg, prefix=''):
     return parser
 
 
+def parse_int_float_bool(val):
+    """Parse string value to be integer, float, or boolean value.
```
Comment (hellock) [inline mmcv/utils/config.py:66]: Add detailed description on:
1. the priority
2. some examples

File: mmcv/engine/utils.py [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+        action=DictAction,
+        help='custom options for evaluation, the key-value pair in xxx=yyy '
+        'format will be kwargs for dataset.evaluate() function')
+
+    return parser
+
+
+def setup_cfg(args, cfg_args):
+    """Set up config.
+
```
Comment (hellock) [inline mmcv/engine/utils.py:120]: Add a detailed description.

**Comment by ZwwWayne** [comment]
As there is a new plan. The proposal of a unified entrypoint has been deprecated and this PR is closed.

**Pull request closed without merging**